hyperspy/learn/mva.py

```diff
@@ -1317,22 +1369,666 @@ def _estimate_elbow_position(self, curve_values):
+            estimate of significant components found in the decomposition step
+            using the elbow method and stored in the 
+            ``learning_results.number_significant_components`` attribute.
+        navigation_mask : boolean numpy array, default : None
+            The navigation locations marked as True are not used in the
+            decomposition.
+        signal_mask : boolean numpy array, default : None
+            The signal locations marked as True are not used in the
+            decomposition.
+        metric : {'elbow','silhouette','gap'} default 'gap'
+            Use distance,silhouette analysis or gap statistics to estimate 
+            the optimal number of clusters.
+            Gap is believed to be, overall, the best metric but it's also 
+            the slowest. Elbow measures the distances between points in 
+            each cluster as an estimate of how well grouped they are and 
+            is the fastest metric. 
+            For elbow the optimal k is the knee or elbow point. 
+            For gap the optimal k is the first k gap(k)>= gap(k+1)-std_error
+            For silhouette the optimal k will be one of the "maxima" found with
+            this method
+        n_ref :  int, default 5
+            Number of random references to use in gap statistics method
+            Gap statistics compares the results from clustering the data to
+            clustering random data. This random clustering is
+            typically averaged n_ref times to get an statistical average
+        cluster_kwargs: dict {}  default empty
+            Additional parameters passed to the clustering algorithm.
+
+        See also
+        --------
+        :py:meth:`~.learn.mva.MVA.plot_cluster_metric`,
+
+        """
+        def distances_within_cluster(data,memberships,squared=True): 
+            return [np.sum(
+                import_sklearn.sklearn.metrics.pairwise.euclidean_distances(
+                data[memberships == c, :]
+                -np.mean(data[memberships == c],axis=0),
+                squared=squared)/
+                (2.*data[memberships == c, :].shape[0])) 
+                for c in np.unique(memberships)]  
+        
+        if use_decomposition_results and number_pca_components is None:
+            number_pca_components = self._get_number_pca_components_for_clustering()
+
+        if max_clusters < 2:
+            raise ValueError("The number of clusters, max_clusters "
+                             "must be specified and be >= 2.")
+
+        to_return = None
+        best_k    = None        
+        k_range   = list(range(1, max_clusters+1))
+        #
+        # for silhouette k starts at 2
+        # for kmeans or gap k starts at 1
+        # As these methods use random numbers we need to
+        # initiate the random number generator to ensure 
+        # consistent/repeatable results
+        #
+        np.random.seed(1)
```

Comment:
Not sure if this is a left over, but I think this seeding should be not be here. It should be set before calling the method in the context of the specific usage.

Reply:
removed - was trying to fix plot problem by fixing seeds